addons/common/functions/fnc_switchPersistentLaser.sqf

```diff
@@ -17,51 +17,95 @@
 
 params ["_enabled"];
 
+// Reset state
+{
+    ACE_player setVariable [QGVAR(laserEnabled_) + str _x, nil];
+} forEach [0, 1, 2];
+
 if (!_enabled) exitWith {
     if (isNil QGVAR(laserKeyDownEH)) exitWith {};
-    ["KeyDown", GVAR(laserKeyDownEH)] call CBA_fnc_removeDisplayHandler;
+
+    removeUserActionEventHandler ["headlights", "Activate", GVAR(laserKeyDownEH)];
+
     ["loadout", GVAR(laserLoadoutEH)] call CBA_fnc_removePlayerEventHandler;
     ["turret", GVAR(laserTurretEH)] call CBA_fnc_removePlayerEventHandler;
     ["vehicle", GVAR(laserVehicleEH)] call CBA_fnc_removePlayerEventHandler;
     ["weapon", GVAR(laserWeaponEH)] call CBA_fnc_removePlayerEventHandler;
 };
 
-GVAR(laserKeyDownEH) = ["KeyDown", {
-    if !((_this select 1) in actionKeys "headlights") exitWith {false};
-    private _weapon = currentWeapon ACE_player;
+private _fnc_getLightLaserState = {
+    private _currentWeapon = currentWeapon ACE_player;
+
+    if (_currentWeapon == "") exitWith {};
+
+    // Ignore in vehicle except FFV
+    if !(ACE_player call CBA_fnc_canUseWeapon) exitWith {};
+
+    // Light/laser state only changes in the next frame
+    [{
+        ACE_player setVariable [
+            QGVAR(laserEnabled_) + str ([ACE_player, _this] call FUNC(getWeaponIndex)),
```

Comment:
possible if binocs equipped, I don't see the harm, but might just be safer to exit?

```suggestion
        private _index = [ACE_player, _this] call FUNC(getWeaponIndex);
        if (_index == -1) exitWith {};
        ACE_player setVariable [
            QGVAR(laserEnabled_) + str _index,
```